akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1402,80 +1402,125 @@ private[stream] object Collect {
 /**
  * INTERNAL API
  */
-@InternalApi private[akka] final class GroupedWithin[T](val n: Int, val d: FiniteDuration) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
-  require(n > 0, "n must be greater than 0")
-  require(d > Duration.Zero)
+@InternalApi private[akka] final class GroupedWeightedWithin[T](val maxWeight: Long, costFn: T ⇒ Long, val interval: FiniteDuration) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
+  require(maxWeight > 0, "maxWeight must be greater than 0")
+  require(interval > Duration.Zero)
 
   val in = Inlet[T]("in")
   val out = Outlet[immutable.Seq[T]]("out")
 
-  override def initialAttributes = DefaultAttributes.groupedWithin
+  override def initialAttributes = DefaultAttributes.groupedWeightedWithin
 
   val shape = FlowShape(in, out)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) with InHandler with OutHandler {
 
     private val buf: VectorBuilder[T] = new VectorBuilder
+    private var pending: T = null.asInstanceOf[T]
+    private var pendingWeight: Long = 0L
     // True if:
     // - buf is nonEmpty
     //       AND
-    // - timer fired OR group is full
-    private var groupClosed = false
+    // - (timer fired
+    //        OR
+    //    totalWeight == maxWeight
+    //        OR
+    //    pending != null
+    //        OR
+    //    upstream completed)
+    private var pushEagerly = false
     private var groupEmitted = true
     private var finished = false
-    private var elements = 0
+    private var totalWeight = 0L
 
-    private val GroupedWithinTimer = "GroupedWithinTimer"
+    private val GroupedWeightedWithinTimer = "GroupedWeightedWithinTimer"
```

Comment:
This could be put in a companion for GroupedWeighedWithin, or in the surrounding stage rather than be a field in every logic instance